torch/nn/utils/spectral_norm.py

```diff
@@ -17,90 +25,168 @@ def __init__(self, name='weight', n_power_iterations=1, dim=0, eps=1e-12):
         weight = module._parameters[name]
-        height = weight.size(dim)
 
-        u = normalize(weight.new_empty(height).normal_(0, 1), dim=0, eps=fn.eps)
+        with torch.no_grad():
+            weight_mat = fn.reshape_weight_to_matrix(weight)
+
+            h, w = weight_mat.size()
+            u = normalize(weight.new_empty(h).normal_(0, 1), dim=0, eps=fn.eps)
+            # compute `v` to satisfy the invariant (see top of this class)
+            # make `u @ W @ v` equal to 1 so at beginning this gives same result
+            v = fn._solve_v_and_rescale(weight_mat, u, target_sigma=1)
+
         delattr(module, fn.name)
         module.register_parameter(fn.name + "_orig", weight)
         # We still need to assign weight back as fn.name because all sorts of
         # things may assume that it exists, e.g., when initializing weights.
         # However, we can't directly assign as it could be an nn.Parameter and
-        # gets added as a parameter. Instead, we register weight.data as a
-        # buffer, which will cause weight to be included in the state dict
-        # and also supports nn.init due to shared storage.
-        module.register_buffer(fn.name, weight.data)
+        # gets added as a parameter. Instead, we register weight.data as a plain
+        # attribute.
+        setattr(module, fn.name, weight.data)
         module.register_buffer(fn.name + "_u", u)
+        module.register_buffer(fn.name + "_v", v)
 
         module.register_forward_pre_hook(fn)
+
+        module._register_state_dict_hook(SpectralNormStateDictHook(fn))
+        module._register_load_state_dict_pre_hook(SpectralNormLoadStateDictPreHook(fn))
         return fn
 
 
+# This is a top level class because Py2 pickle doesn't like inner class nor an
+# instancemethod.
+class SpectralNormLoadStateDictPreHook(object):
+    # See docstring of SpectralNorm._version on the changes to spectral_norm.
+    def __init__(self, fn):
+        self.fn = fn
+
+    # For state_dict with version None, we have
+    #
+    #    u = normalize(W_orig @ v)
+    #    W = W_orig / sigma, where sigma = u @ W_orig @ v
+    #
+    # To compute `v`, we solve `W_orig @ x = u`, and let
+    #    v = x / (u @ W_orig @ x) * (W / W_orig).
+    def __call__(self, state_dict, prefix, local_metadata, strict,
+                 missing_keys, unexpected_keys, error_msgs):
+        fn = self.fn
+        version = local_metadata.get('spectral_norm', {}).get(fn.name + '.version', None)
+        if version is None or version < 1:
+            with torch.no_grad():
+                weight_orig = state_dict[prefix + fn.name + '_orig']
+                weight = state_dict.pop(prefix + fn.name)
+                sigma = (weight_orig / weight).mean()
+                weight_mat = fn.reshape_weight_to_matrix(weight_orig)
+                u = state_dict[prefix + fn.name + '_u']
+                v = fn._solve_v_and_rescale(weight_mat, u, sigma)
```

Comment:
Is it OK to have computation during loading the state dict? I think it probably is, but it should be a conscious decision.

Reply:
Yes it is intended. Old state dict don't have `v` so we have to "solve" for it somehow. It can't be after loading the state dict because we lose info about the version in the state dict by then.